### DISCUSSION TITLE
Staking project specification

### DISCUSSION THREAD

**Discussion by henry-hz** [blockchain-audit/labs#59] [Ideas]
We will choose one or more specs that make sense. Check if your specs are providing one of these mechanisms.

Supose our token is equal 1 ILS, the daily interest rate in Israel is approximately:
```
4/365 = 0.0109589
```
Make sure that your daily interest rate is bigger than the free risk gov.

## 10 DeFi Staking Mechanisms Across Different Fields:

1. **Security:**  **Validator Staking:** Users stake tokens to become validators who verify transactions and secure the network. They earn rewards based on the amount staked and the network activity. (e.g. Ethereum 2.0)
2. **Lending/Borrowing:**  **Liquidity Staking:** Users stake tokens to provide liquidity to lending pools. Borrowers can access loans from these pools, and stakers earn interest based on the utilization rate. (e.g. Compound, Aave)
3. **Insurance:**  **Coverage Staking:** Users stake tokens to act as a decentralized insurance pool. When a protocol suffers a loss, staked tokens are used to compensate victims. Stakers earn rewards based on the amount staked and the overall risk. (e.g. Nexus Mutual, DIA)
4. **Gaming:**  **Play-to-Earn Staking:** Users stake tokens to earn in-game rewards or access exclusive features within a game. Staking can also be used for governance purposes, allowing players to vote on game development decisions. (e.g. Axie Infinity, The Sandbox)
5. **Prediction Markets:**  **Staking for Accuracy:** Users stake tokens to predict the outcome of real-world events. Those with accurate predictions earn rewards from a pool funded by incorrect guesses. (e.g. Augur, Gnosis)
6. **Content Creation:**  **Content Curation Staking:** Users stake tokens to curate content on a platform. Staking power determines the visibility of content, and users with high-quality curation earn rewards. (e.g. Livepeer, Theta Network)
7. **Identity & Reputation:**  **Reputation Staking:** Users stake tokens to build a verifiable reputation score on a decentralized identity platform. Higher reputation scores unlock access to services or discounts. (e.g. Civic, Sovrin)
8. **Decentralized Storage:**  **Storage Provider Staking:** Users stake tokens to become storage providers on a decentralized storage network. They earn rewards based on the amount of data they store and the network traffic. (e.g. Filecoin, Siacoin)
9. **Decentralized Autonomous Organizations (DAOs):**  **Governance Staking:**  DAO members stake tokens to vote on proposals and influence the direction of the organization. Staking weight often increases with the amount staked. (e.g. MakerDAO, Uniswap)
10. **Yield Farming:**  **Short-term Liquidity Staking:** Users strategically move their tokens between different DeFi protocols to capture the highest possible returns. This is a complex strategy with high risks  and requires active management.

## Examples of staking projects that are different from the native ETH:

Here are 10 examples of staking projects that are not connected to the native coin, like Lido, but allow staking for other reasons:

1. **Kleros (KLR)**: Kleros is a decentralized jury network that allows users to stake KLR tokens to participate in dispute resolution. Stakers earn rewards based on their participation and accuracy in resolving disputes.

2. **Livepeer (LPT)**: Livepeer is a decentralized video streaming network that allows users to stake LPT tokens to contribute bandwidth and transcoding resources to the network. Stakers earn rewards based on the amount of resources they contribute.

3. **The [Graph](https://thegraph.com/) (GRT)**: The Graph is a decentralized indexing protocol that allows users to stake GRT tokens to curate and serve data to applications. Stakers earn rewards based on the amount of data they curate and the quality of the service they provide.

4. **[Ocean](https://oceanprotocol.com/) Protocol (OCEAN)**: Ocean Protocol is a decentralized data marketplace that allows users to stake OCEAN tokens to provide and access data on the network. Stakers earn rewards based on the amount of data they provide or access.

5. **dYdX (DYDX)**: dYdX is a decentralized derivatives exchange that allows users to stake DYDX tokens to earn trading fee discounts and participate in governance decisions.

6. **Augur (REP)**: Augur is a decentralized prediction market that allows users to stake REP tokens to report on the outcomes of events. Stakers earn rewards based on the accuracy of their reports.

7. **Band Protocol (BAND)**: Band Protocol is a decentralized oracle network that allows users to stake BAND tokens to provide secure and reliable data feeds to smart contracts. Stakers earn rewards based on the amount of data they provide and the quality of the service they offer.

8. **NuCypher (NU)**: NuCypher is a decentralized key management system that allows users to stake NU tokens to provide storage and access control for encrypted data. Stakers earn rewards based on the amount of data they store and the security they provide.

9. **Nexus Mutual (NXM)**: Nexus Mutual is a decentralized insurance platform that allows users to stake NXM tokens to provide and purchase parametric insurance. Stakers earn rewards based on the amount of coverage they provide and the premiums they collect.

10. **Storj (STORJ)**: Storj is a decentralized cloud storage network that allows users to stake STORJ tokens to provide and access storage space on the network. Stakers earn rewards based on the amount of storage they provide and the bandwidth they contribute.

**Comment by henry-hz** [comment]
There are million tokens in the pool. 
Each depositor can withdraw his tokens starting 7 days from the day of the deposit.
When the tokens are deposited, the calculation is like this:
the amount of tokens deposited *100 / the amount of tokens in the pool after the deposit.
For example: Nitai deposited 300000 tokens in a pool that contained a million tokens. 
After 7 days from the day of the deposit, Nitai will be able to withdraw 23.0769230769% of the pool.

**Reply by henry-hz** [reply]
There is one point here to change, you can add a cap multiplier, like always divide 2% of the pool

**Comment by henry-hz** [comment]
By: Ruth Klirs, Rachel Paley,  Yael Taitelbom
REWARDS-4-7 is a smart contract that manages a pool of 1,000,000 cloud tokens. Participants can embed their cloud tokens in the contract, which will be stored in a variable called total deposits
. To be eligible for rewards, participants must hold their tokens in stake for at least 7 days.

The rewards are distributed proportionally based on the amount of the bet that the user wishes to withdraw in relation to the total deposits of the participants that are currently in place.
The reward is 2% of the pool of reward tokens multiplied by the ratio of the amount withdrawn to the amount of tokens in total deposits.
if the total-deposits is bigger than Reward-pool So we will add new tokens for Reward-pool.

Below is a breakdown of how REWARDS-4-7 works

example
Scenario 1: Deposits made on the same day by different people:
Cloud storage with 1,000,000 cloud tokens.
Investors:
Staker 1: deposits 10,000 tokens on day 1 (total deposits: 10,000)
Staker 2: convert 5000 tokens on day 1 (total deposits: 15,000)

Withdrawal of deposits on the 8th:

Staker 1: withdraws 2,000 tokens and receives a reward of: ((2000÷15000)×2÷100×1000000=2666)
status:
    • total deposits: 13,000
    • Reward pool: 1000000- 2666=997333

Staker 2: withdraws 2000 tokens and receives a reward of:
997334×(2÷100)×(2000÷13000)=3068
status:
    • total deposits: 11,000
    • Reward pool: 997333-3068=994265



main points:
Participants are rewarded proportionally to their contribution to the total bet
Betting earlier increases the chance of getting a bigger reward because the total bet amount may be lower.
Betting after a 7 day window disqualifies users from receiving rewards for that particular round.
This structure incentivizes participants to stake their tokens early and hold them for a specified period to maximize their potential rewards from a collective staking contract.

**Comment by henry-hz** [comment]
[Meira Yavetz]
Example:
Reuven deposits a token in type-MyToken.
Reuven can always deposit, every day and every hour
On the day Reuven wants to withdraw the money, a check will be made- to see if he can receive a bonus.

Calculation of the test:

Every person who deposits a token-
the amount goes into a pool.
The bonus is received according to the ratio between the amount of 
deposits that were in the pool minimum 7 days to the current balance in the pool.

Reuven receives:
1. The amount he withdrew from how much he deposited in the deposit pool.
2. If the depositor's amount meets the calculation and he needs to receive a bonus, 
he gets it from the big pool - the bonus pool.
3. The percentage he will receive from the large pool is not from the entire pool 
but from 10 percent of the pool so that it remains for additional investors

**Comment by henry-hz** [comment]
[Elisheva Nankensky]
A countract owns 1000000 cloud tokens. Anyone could stakes cloud token into the contract. The staked amount is locked for 7 days (if you add staking, the 7 days starts from the last day you staked).
When you withdraw you get your original staked amount + a reward(the reward is constantly 10000000) propotional to your portion of the total stake at that time.
exaample:
total staked amout=100000
Staker 1: Stakes 5,000 tokens on day 2 (beginDate 2)   ||    total staked amout = 105000
Staker 2: Stakes 10,000 tokens on day 3 (beginDate 3)  ||    total staked amout = 115000
Staker 3: Stakes 7,000 tokens on day 5 (beginDate 5)   ||    total staked amout = 122000


Staker 1: After 7 days (on day 9), calculates a reward of 40,983 tokens (assuming total stake remains 122,000) - [(5,000 * 1,000,000) / 122,000]
Staker 2: After 9 days (on day 12), calculates a reward of 8,547 tokens (assuming total stake remains 117,000) - [(10,000 * 1,000,000) / 117,000]
Staker 3: After 7 days (on day 12), calculates a reward of 65,420 tokens (assuming total stake remains 107,000) - [(7,000 * 1,000,000) / 107,000]

**Comment by henry-hz** [comment]
[Sari Kahan]
explanation :
    in my totalsupply staking have in begining 1000000 MyToken 
    and everyone can to stake and it is possible to out this after 7 days
    everyone get percent reward from 1000000 MyToken in accordance to her stake percent from all staker

exampels :
    in one day : alice stake 5000 MyToken
    in 2 day : bob stake 1000 MyToken
    in 3 day : adam stake 10000 MyToken

    in 7 day : alice want to withdraw his stake + reward, he get 5000(his stake) + 200000(reward) - 
    1000000 / (25000 / 5000)

    in 8 day : alice want to withdraw his stake + reward, he get 1000(his stake) + 400000(reward) - 
    800000 / (2000 / 1000)

**Reply by tova100** [reply]
I like your idea 
you:
tokens/(total/user's token)
in the calculation i think change :
tokens/(user's token/total)
Advantage:
He receives a reward but in proportion to what he deposited (smaller than what exists).
and if we want to reduce more, we can double by some percentage of the tokens.

**Comment by henry-hz** [comment]
[Yehudis David]
this staking specs:
from the day the contract is deployed - any one can stake his tokens on the contract and recieve rewards
the staker has to lock his tokens for atlist 7 days - after 7 days he can unlock his tokens and recieve his rewared
the rewared is calculated like this: his percent of staking from the totalstaking * the total rewared




stakers:
* staker 1 : stakes 5000 tokens on any day - his begin-date starts and his tokens are locked for 7 days (day 12).
* staker 2 : stakes 20000 tokens on any day - his begin-date starts and his tokens are locked for 7 days (day 15).
* staker 3 : stakes 6000 tokens on any day - his begin-date starts and his tokens are locked for 7 days (day 21).
* staker 4 : stakes 100000 tokens on any day - his begin-date starts and his tokens are locked for 7 days (day 27).

rewards:
* staker 1 : after 7 days (on day 18) he unlockes his tokens and gets a rewared 200000 (the totale staking 25000)
[(5000/25000)*1000000]
* staker 2 :decides to unlock his tokens 12 days after (on day 27) and gets a rewared 769230 (the totale staking 25000)
[(20000/26000)*1000000]
* staker 3 :decides to unlock his tokens after 7 days (on day 28) and gets a rewared 56603 (the total staking 106000)
[(6000/100000)*1000000]

**Comment by henry-hz** [comment]
[Wolpin]
There are million tokens in the pool.
Each depositor can withdraw his tokens starting 7 days from the day of the deposit.
When the tokens are deposited, the calculation is like this:
the amount of tokens deposited *100 / the amount of tokens in the pool after the deposit.
For example: Nitai deposited 300000 tokens in a pool that contained a million tokens.
After 7 days from the day of the deposit, Nitai will be able to withdraw 23.0769230769% of the pool.

**Reply by henry-hz** [reply]
After 7 days we withdraw 23% of the pool, so how the 1 million will be able to achieve its objective ?  [see the 10 objectives]
